### PR TITLE
selfdrive/car: ban cereal and capnp

### PR DESCRIPTION
**Pull request by sshane**
instead of https://github.com/commaai/openpilot/pull/33197

while thinking about it, I realized we might not need to switch to a full blown class in the first iteration. will try to reduce touching things, and just convert log.CanData to a custom dataclass with generic send/receive functions passed into fingerprinting/car interface code

### CODE REVIEW

File: selfdrive/car/data_structures.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  body = auto()  # body control module
+
+  # Toyota only
+  dsu = auto()
+
+  # Honda only
+  vsa = auto()  # Vehicle Stability Assist
+  programmedFuelInjection = auto()
+
+  debug = auto()
```
Comment (sshane) [inline selfdrive/car/data_structures.py:34]: @adeebshihadeh do you see this duplication as okay?

Comment (sshane) [reply selfdrive/car/data_structures.py:34]: We can use a metaclass to achieve the capnp builder behavior, right now we'd have to specify defaults in the dataclass or build the object all at once

Comment (adeebshihadeh) [reply selfdrive/car/data_structures.py]: yeah, this is fine

**Comment by sshane** [comment]
We have builder, automatic defaults, and proper static analysis typing now! Operates similarly to Enum's `auto()` object.

```python

@auto_dataclass
class TestClass:
  a: int = auto_field()
  b: float = auto_field()
  c: str = auto_field()
  d: bytes = auto_field()
  e: list[int] = auto_field()
  f: tuple[int] = auto_field()
  g: set[int] = auto_field()
  h: dict[str, int] = auto_field()
  i: bool = auto_field()
  ecu: CarParams.Ecu = auto_field()
  carFw: CarParams.CarFw = auto_field()
  carFwList: list[CarParams.CarFw] = auto_field()

TestClass()
Out[3]: TestClass(a=0, b=0.0, c='', d=b'', e=[], f=(), g=set(), h={}, i=False, ecu=<Ecu.eps: 'eps'>, carFw=CarParams.CarFw(ecu=<Ecu.unknown: 'unknown'>, fwVersion=b'', address=0, subAddress=0, responseAddress=0, request=[], brand='', bus=0, logging=False, obdMultiplexing=False), carFwList=[])
```

**Comment by sshane** [comment]
The docs diff is acceptable:

```python
CP_cereal = car.CarParams()
CP_cereal.minEnableSpeed = 25.5 * CV.MPH_TO_MS
print(CP_cereal.minEnableSpeed * CV.MS_TO_MPH)

CP = CarParams()
CP.minEnableSpeed = 25.5 * CV.MPH_TO_MS
print(CP.minEnableSpeed * CV.MS_TO_MPH)
25.49999982182606
25.500000000000004
```

**Comment by sshane** [comment]
@adeebshihadeh what's left is CarControl (which should be pretty easy) and the events (biggest blocker). How do you propose we handle car-specific behavior like this?

make `accurate_steer_angle_seen` generic?
https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/toyota/interface.py#L159-L160

hardcode this `notCar` behavior?
https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/body/interface.py#L30-L36

hardcode the standstill + brake fault prevention logic?
https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/gm/interface.py#L207-L229

a lot of cars implement different offsets for the alerts, make standard?
https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/chrysler/interface.py#L87-L93

custom alert for disengaging unexpectedly
https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/honda/interface.py#L238-L251

I think possibly keeping brand-specific events might be a good idea for the time being, but we'd need to duplicate the states as well. I can do either, but removing the events would take a bit longer to standardize some of this stuff.

**Comment by adeebshihadeh** [comment]
> make `accurate_steer_angle_seen` generic?

I'd add `vehicleSensorsInvalid` to carState.

https://github.com/commaai/openpilot/blob/ee62b9c88baf4a53f6db88381da2b8052c3fab17/selfdrive/car/toyota/interface.py#L159-L160
> 
> hardcode this `notCar` behavior?
Yeah, seems fine to just move that to `card` for now.

> hardcode the standstill + brake fault prevention logic?
I don't understand this one?

> a lot of cars implement different offsets for the alerts, make standard?
I'd say yes, but there may be a reason why it's done like this.

> custom alert for disengaging unexpectedly

Not sure about this one, have to think more about it.

**Comment by sshane** [comment]
> I don't understand this one?

The GM camera cars have specific logic to allow engaging under the min engage speed if the foot is on the brake at a stop, with a special brake pressed threshold to avoid another fault where the user isn't pressing the brake hard enough at a stop only. Not at a stop we need a lower threshold to detect brake pressed earlier, so this isn't easily made generic.

By hard-coding logic, are we fine with something like this in card?

```python
if CP.carFingerprint in (CAR...,):
  # interface logic here
```

**Comment by sshane** [comment]
process replay diffs:

A lot are small minute differences due to switching from float32 precision to float64 using Python's float type. Nissan calculates mean of floats to get steering pressed for example. Toyota rate limit being different was because its CarControllerParams didn't call `which` as a function, and the `<torque which-enum>` defines a `__str__` method, or perhaps `__eq__`

**Comment by sshane** [comment]
I've reviewed the car interface code, can you review card.py and structs.py? For the car tests touched here, I am going to leave them in openpilot for now, but I can see splitting them up like I did for the docs test in a follow up PR (not going to do here since they're all so huge and require some more thought).

File: selfdrive/car/card.py [hunk 2]
```diff
@@ -58,8 +61,74 @@ def can_send(msgs: list[CanData]) -> None:
...
+      ret = {}
+      for field in getattr(obj, _FIELDS):  # similar to dataclasses.fields()
+        ret[field] = _asdictref_inner(getattr(obj, field))
+      return ret
+    elif isinstance(obj, (tuple, list)):
+      return type(obj)(_asdictref_inner(v) for v in obj)
+    else:
+      return obj
+
+  return _asdictref_inner(obj)
```
Comment (sshane) [inline selfdrive/car/card.py:88]: We unfortunately have to rewrite these to take advantage of referencing vs. deepcopying and avoiding calling `type()` on every single field recursively, which take around ~7-8% total vs 3-4% with this

**Pull request merged**